python/paddle/v2/fluid/io.py

```diff
@@ -234,6 +261,21 @@ def append_fetch_ops(inference_program,
             attrs={'col': i})
 
 
+def get_parameters(program):
+    parameter_list = []
+    input_args = set()
+    for block in program.blocks:
+        for op in block.ops:
+            if op.desc.type() != 'feed':
+                input_args.update(op.desc.input_arg_names())
+
+    for var in program.list_vars():
+        if is_persistable(var) and var.name in input_args:
+            parameter_list.append(var)
```

Comment:
It is actually not a parameter list but a persistable variable list. Normally, the program should not contain unreferenced variables, so `if var.name in input_args` should be removed. When loading, if a persistable variable is absent, there should be some error message.

Reply:
Thanks for the explanation. Then I guess we don't need to define this function. There is already a save_persistable method, so I will use that one instead.

Reply:
But I think we still need to exclude 'feed' and 'fetch' variables right (because they have been added to the program desc)? They are also persistable and we don't want to store them.

Reply:
I tried to remove `var.name in input_args`, then 'feed' and 'fetch' variables are also included, which incurs an error "The type of var fetch is unsupported" since the type of feed/fetch is `vector<lodTensor>` and is not supported by  load / save op.

Reply:
I think this implementation can potentially solve the problem described in PR #8020

Reply:
I will try including these changes to actually verify.

Reply:
>  But I think we still need to exclude 'feed' and 'fetch' variables right

@kexinzhao  Can we change the function `is_persistable()` or rewrite `load/save_persistables()` to exclude `feed` and `fetch` variables? 

> I think this implementation can potentially solve the problem described in PR #8020

@sidgoyal78 I think the problem in #8020 is that, `means` of bn is not parameter but persistable variable. In fact, we should save all persistable variables in `save_inference_model`, not only parameters. I think about this for a long time, and there are some issues for this: #7931 #7163

Reply:
Yeah. 👍

Reply:
> Can we change the function is_persistable() or rewrite load/save_persistables() to exclude feed and fetch variables?

I think of three ways to redefine load/save_persistables():
1. One is like this:
```python
def load_persitables(xxx):
     parameter_list = get_parameters(program)
     load_var(xxx, parameter_list)
```
which basically moved the usage of `get_parameters` inside `save_load_persistables`.
You can also get rid of the `get_parameters` method, move its code directly into save/load_persistable, but we prefer not to do this because of duplicate code in two functions.

2. If we don't want to use the code in `get_parameters()` method, then basically we want to modify the `is_persistable(var)` predicate function so that it can exclude 'feed' and 'fetch' vars.
Note that since this is a predicate, we only have the variable as input. 
Although as show in framework.py that `class Variable` has a data member `op`, this `op` will only be set to the operator that output this variable. Meaning that if a variable is not an output of any operator (e.g., `feed` and weights parameters), this var.op == None. 

So we cannot use code like below to exclude 'feed' and 'fetch'
```python
def is_persistable(var):
    if var.op.desc.type() == 'feed' or var.op.desc.type() == 'fetch':
        return false
    return var.persistable
```

3.  Just like we define the feed/fetch operator type to be fixed as 'feed'/'fetch', we can also fix the name of the feed/fetch variable to be 'feed'/'fetch' (or some better names). This means that we can get rid of the API in the current Inference Design that optionally allows user to provide its own 'feed_holder_name' and 'fetch_holder_name'.   For this design, we can simply modify the `is_persistable` as follows:
```python
def is_persistable(var):
    if var.desc.name() == 'feed' or var.desc.name() == 'fetch':
        return false
    return var.persistable
```
If we want to go with this option, we can firstly do a quick fix in this pr using the code above. Then fix the feed/fetch var name, modify API accordingly, set some global const kFeedVarName in C++ and pybind it to python, etc in the future PR.

@Xreki @sidgoyal78,  which option do your prefer or do you have other suggestions?

Reply:
I'd like the 2nd method. But I am not sure whether it is suitable to use `var.op.desc.type()`. In the C++ definition of `VarDesc`, there is not a member to record the belonged op. Also, a variable may be shared among multiple op.

For the 3rd method. I think it is not suitable to use the name, but may be we can use the type, which should be `FEED_MINIBATCH`.